### PR TITLE
Feature/save data hook

### PR DESCRIPTION
**Pull request by anton-karlovskiy**
1. Save Data Hook added
2. Save Data Hook based demo project added
3. demo project README, repository README, and hook/README are updated
4. build configuration updated and deployed again with Client Hint version
4. Client Hint based demo project updated in naming

### CODE REVIEW

**Comment by anton-karlovskiy** [comment]
@addyosmani
I think this project is sibling project with the previous Client Hint version.
So please review the project naming as well.

I think we should update the projects naming according to https://github.com/GoogleChromeLabs/adaptive-loading/issues/65

File: hooks/save-data/index.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+const UNSUPPORT_MESSAGE = 'The Save Data API is not supported on this platform.';
+const SAVE_DATA_MODE = {
+  ON: 'on',
+  OFF: 'off'
+};
+
+const useSaveData = () => {
+  let initialSaveData;
+  if ('connection' in navigator) {
+    initialSaveData = navigator.connection.saveData ? SAVE_DATA_MODE.ON : SAVE_DATA_MODE.OFF;
```
Comment (addyosmani) [inline hooks/save-data/index.js]: `navigator.connection.saveData` is a boolean value, but here we are changing it to mean a string:

```js
const SAVE_DATA_MODE = {
  ON: 'on',
  OFF: 'off'
};
```

I think we should stick with treating it as a true/false value when passing along to `useState`. Would that work?

File: hooks/save-data/index.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+  return {saveData, setSaveData};
+};
+
+export {
+  useSaveData,
+  // exported for Unit Tests
+  UNSUPPORT_MESSAGE,
+  SAVE_DATA_MODE
+};
```
Comment (addyosmani) [inline hooks/save-data/index.js]: Similar to the comments on the other thread, I think we should refactor this to only pass along the minimal useful information and move unit-test specific values to test configuration.

File: hooks/save-data/index.js (hunk 1 above)
Comment (anton-karlovskiy) [reply hooks/save-data/index.js]: Of course, let me update it that way.

File: hooks/save-data/index.js (hunk 2 above)
Comment (anton-karlovskiy) [reply hooks/save-data/index.js]: Yes, of course. It's in progress in the other branch. No worries. :)

File: README.md [hunk 3]
```diff
@@ -25,7 +25,8 @@ This repo contains several different pieces for the Adaptive Loading project: Re
 
 ### Data Saver-style patterns
 
-* [React Twitter - save-data loading](https://github.com/GoogleChromeLabs/adaptive-loading/tree/master/react-twitter-save-data-loading) ([demo](https://adaptive-loading.web.app/react-twitter-save-data-loading/))
+* [React Twitter - save-data loading based on Client Hint](https://github.com/GoogleChromeLabs/adaptive-loading/tree/master/react-twitter-save-data-loading(client-hint)) ([demo](https://adaptive-loading.web.app/react-twitter-save-data-loading(client-hint)/))
+* [React Twitter - save-data loading based on Hook](https://github.com/GoogleChromeLabs/adaptive-loading/tree/master/react-twitter-save-data-loading(hook)) ([demo](https://adaptive-loading.web.app/react-twitter-save-data-loading(hook)/))
```
Comment (addyosmani) [inline README.md]: Great work on this demo, @anton-karlovskiy!

Comment (anton-karlovskiy) [reply README.md]: Thank you.

**Comment by addyosmani** [comment]
For my knowledge, will this PR be getting updated or are you aiming to PR all the changes based on the values feedback at the same time?

**Comment by anton-karlovskiy** [comment]
> For my knowledge, will this PR be getting updated or are you aiming to PR all the changes based on the values feedback at the same time?

I think we can merge all current PRs based on your comments then we can branch out for hook update like returning values and memory hook.
In this way, we could reduce duplicated works I guess.
So for now, my priority is to handle all your comments so we can land on PRs.

**Comment by addyosmani** [comment]
Cool. Thank you. Once the conflicts on this one have been resolved we should be able to merge.

**Comment by anton-karlovskiy** [comment]
@addyosmani 
This is ready now.
All comments and relevant updates are done.

**Comment by addyosmani** [comment]
Would you mind resolving the conflicts just one more time?

**Comment by anton-karlovskiy** [comment]
> Would you mind resolving the conflicts just one more time?

Sure thing!

**Comment by anton-karlovskiy** [comment]
> Would you mind resolving the conflicts just one more time?

Absolutely! It's done. Please review it.

**Pull request merged**